sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java

```diff
@@ -36,53 +45,101 @@
     private final ReactorDispatcher dispatcher;
     private final Disposable subscriptions;
     private final AtomicBoolean isDisposed = new AtomicBoolean();
+    private final Sinks.Empty<Void> isClosedMono = Sinks.empty();
     private final Flux<Message> messagesProcessor;
+    private final AmqpRetryOptions retryOptions;
     private final ClientLogger logger = new ClientLogger(ReactorReceiver.class);
     private final Flux<AmqpEndpointState> endpointStates;
 
     private final AtomicReference<Supplier<Integer>> creditSupplier = new AtomicReference<>();
 
-    protected ReactorReceiver(String entityPath, Receiver receiver, ReceiveLinkHandler handler,
-        TokenManager tokenManager, ReactorDispatcher dispatcher) {
+    protected ReactorReceiver(AmqpConnection amqpConnection, String entityPath, Receiver receiver,
+        ReceiveLinkHandler handler, TokenManager tokenManager, ReactorDispatcher dispatcher,
+        AmqpRetryOptions retryOptions) {
         this.entityPath = entityPath;
         this.receiver = receiver;
         this.handler = handler;
         this.tokenManager = tokenManager;
         this.dispatcher = dispatcher;
         this.messagesProcessor = this.handler.getDeliveredMessages()
-            .map(this::decodeDelivery)
-            .doOnNext(next -> {
-                if (receiver.getRemoteCredit() == 0 && !isDisposed.get()) {
-                    final Supplier<Integer> supplier = creditSupplier.get();
-                    if (supplier == null) {
-                        return;
+            .flatMap(delivery -> {
```

Comment:
We check the remoteCredits in the dispatcher and get credits to add if remoteCredits are 0.

RemoteCredits is a client-side calculation based on:
`remoteCredits = creditsOnLink - pendingDeliveries`